core/iwasm/fast-jit/fe/jit_emit_memory.c

```diff
@@ -5,18 +5,131 @@
 
 #include "jit_emit_memory.h"
 #include "../jit_frontend.h"
+#include "fe/jit_emit_exception.h"
+
+static JitReg
+check_and_seek(JitCompContext *cc, JitReg addr, JitReg offset, uint32 bytes)
+{
+    JitReg addrL, offset1, memory_boundary, memory_data, maddr;
+    /* the default memory */
+    uint32 mem_idx = 0;
+
+    /* check */
+    /* addrL = (int64_t)addr */
+    addrL = jit_cc_new_reg_I64(cc);
+    GEN_INSN(I32TOI64, addrL, addr);
+
+    /* offset1 = offset + addrL */
+    offset1 = jit_cc_new_reg_I64(cc);
+    GEN_INSN(ADD, offset1, offset, addrL);
+
+    /* if (offset1  >= memory_boundary) goto EXCEPTION */
+    switch (bytes) {
+        case 1:
+        {
+            memory_boundary =
+                get_mem_bound_check_1byte_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 2:
+        {
+            memory_boundary =
+                get_mem_bound_check_2bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 4:
+        {
+            memory_boundary =
+                get_mem_bound_check_4bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 8:
+        {
+            memory_boundary =
+                get_mem_bound_check_8bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        case 16:
+        {
+            memory_boundary =
+                get_mem_bound_check_16bytes_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+        default:
+        {
+            memory_boundary = get_memory_data_end_reg(cc->jit_frame, mem_idx);
+            break;
+        }
+    }
+
+    GEN_INSN(CMP, cc->cmp_reg, offset1, memory_boundary);
+    if (!jit_emit_exception(cc, EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS, JIT_OP_BGEU,
+                            cc->cmp_reg, 0)) {
+        goto fail;
+    }
+
+    /* seek */
+    memory_data = get_memory_data_reg(cc->jit_frame, mem_idx);
+    maddr = jit_cc_new_reg_I64(cc);
+    GEN_INSN(ADD, maddr, memory_data, offset1);
+    return maddr;
+fail:
+    return 0;
+}
 
 bool
 jit_compile_op_i32_load(JitCompContext *cc, uint32 align, uint32 offset,
                         uint32 bytes, bool sign, bool atomic)
 {
+    JitReg addr, maddr, value;
+
+    POP_I32(addr);
+
+    maddr =
+        check_and_seek(cc, addr, NEW_CONST(I64, offset), NEW_CONST(I64, bytes));
+    if (!maddr) {
+        goto fail;
+    }
+
+    value = jit_cc_new_reg_I32(cc);
+    if (sign) {
+        GEN_INSN(LDI32, value, maddr, NEW_CONST(I32, 0));
```

Comment:
`bytes` argument is ignored, should use `LDI32`, `LDI16`, `LDU16`, `LDI8`, `LDU8`